src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,303 @@
+        {
+            var options = syntaxContext.Options;
+            var syntaxTree = syntaxContext.Node.SyntaxTree;
+            var cancellationToken = syntaxContext.CancellationToken;
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var styleOption = optionSet.GetOption(CSharpCodeStyleOptions.PreferLocalOverAnonymousFunction);
+            if (!styleOption.Value)
+            {
+                // Bail immediately if the user has disabled this feature.
+                return;
+            }
+
+            // Local functions are only available in C# 7.0 and above.  Don't offer this refactoring
+            // in projects targetting a lesser version.
+            if (((CSharpParseOptions)syntaxTree.Options).LanguageVersion < LanguageVersion.CSharp7)
+            {
+                return;
+            }
+
+            var severity = styleOption.Notification.Value;
+            var anonymousFunction = (AnonymousFunctionExpressionSyntax)syntaxContext.Node;
+
+            var semanticModel = syntaxContext.SemanticModel;
+            if (!CheckForPattern(semanticModel, anonymousFunction, cancellationToken,
+                    out var localDeclaration))
+            {
+                return;
+            }
+
+            if (localDeclaration.Declaration.Variables.Count != 1)
+            {
+                return;
+            }
+
+            if (!(localDeclaration.Parent is BlockSyntax block))
+            {
+                return;
+            }
+
+            var local = semanticModel.GetDeclaredSymbol(localDeclaration.Declaration.Variables[0], cancellationToken);
+            if (local == null)
+            {
+                return;
+            }
+
+            if (IsWrittenAfter(semanticModel, local, block, anonymousFunction, cancellationToken))
+            {
+                return;
+            }
+
+            var delegateType = semanticModel.GetTypeInfo(anonymousFunction, cancellationToken).ConvertedType as INamedTypeSymbol;
+            if (!delegateType.IsDelegateType() ||
+                delegateType.DelegateInvokeMethod == null)
+            {
+                return;
+            }
+
+            // Looks good!
+            var additionalLocations = ImmutableArray.Create(
+                localDeclaration.GetLocation(),
+                anonymousFunction.GetLocation());
+
+            if (severity != DiagnosticSeverity.Hidden)
+            {
+                // If the diagnostic is not hidden, then just place the user visible part
+                // on the local being initialized with the lambda.
+                syntaxContext.ReportDiagnostic(Diagnostic.Create(
+                    GetDescriptorWithSeverity(severity),
+                    localDeclaration.Declaration.Variables[0].Identifier.GetLocation(),
```

Comment:
Haven't we been putting most lambda diagnostics on `=>`  Should we do the same here?